### PR TITLE
Dc/taco 425 adding route reloader and dynamic route config

### PR DESCRIPTION
**Pull request by khappucino**
This will allow the RouteReloader to ingest a simple JSON file instead of a HOCON with substitution references.  The output will be a list of POJOS that upstream sources can use to template route/client configs. 
[
  {
      "name": "route1",
      "path": "/path1/",
      "client_name": "client1",
      "port_number": 1234,
      "tls_enabled": false,
      "ip_addresses": ["1.2.3.4", "1.2.3.5"]
  },
  {
      "name": "route2",
      "path": "/path2/",
      "client_name": "client2",
      "port_number": 5678,
      "tls_enabled": true,
      "ip_addresses": ["2.2.3.4", "2.2.3.5"]
  },
  {
      "name": "route3",
      "path": "/path3/",
      "client_name": "client3",
      "port_number": 9012,
      "tls_enabled": false,
      "ip_addresses": []
  }
]

### CODE REVIEW

File: xio-core/src/main/java/com/xjeffrose/xio/config/DynamicClientConfig.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  public int getPort() {
+    return port;
+  }
+
+  public boolean isTlsEnabled() {
+    return tlsEnabled;
+  }
+
+  public boolean equals(Object other) {
```
Comment (Rivukis) [inline xio-core/src/main/java/com/xjeffrose/xio/config/DynamicClientConfig.java]: There's a Lombok helper for equals and hashcode. Can put `@EqualsAndHashCode` above the class name

File: xio-core/src/main/java/com/xjeffrose/xio/config/DynamicRouteConfig.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+  public String getPath() {
+    return path;
+  }
+
+  public ArrayList<DynamicClientConfig> getClientConfigs() {
+    return clientConfigs;
+  }
+
+  public boolean equals(Object other) {
```
Comment (Rivukis) [inline xio-core/src/main/java/com/xjeffrose/xio/config/DynamicRouteConfig.java]: Same Lombok helper here as well

File: xio-core/build.gradle [hunk 3]
```diff
@@ -27,6 +27,7 @@ group = project_group
 apply from: '../alpn-boot.gradle'
 
 dependencies {
+  implementation group: 'com.google.code.gson', name: 'gson', version: '2.8.0'
```
Comment (manimaul) [inline xio-core/build.gradle]: we're preferring Jackson over Gson

File: xio-core/src/main/java/com/xjeffrose/xio/config/DynamicClientConfig.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
+package com.xjeffrose.xio.config;
+
+import java.util.Objects;
+
+/** This class is a POJO representation of the input data that maps to a ClientConfig */
+public class DynamicClientConfig {
```
Comment (manimaul) [inline xio-core/src/main/java/com/xjeffrose/xio/config/DynamicClientConfig.java]: lombok please

File: xio-core/src/main/java/com/xjeffrose/xio/config/DynamicClientConfig.java (hunk 1 above)
Comment (manimaul) [inline xio-core/src/main/java/com/xjeffrose/xio/config/DynamicClientConfig.java]: as @Rivukis commented earlier - https://projectlombok.org/features/EqualsAndHashCode

File: xio-core/src/main/java/com/xjeffrose/xio/config/DynamicRouteConfig.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
+package com.xjeffrose.xio.config;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Objects;
+
+/** This class is the POJO representation of the input data that maps to a RouteConfig/RouteState */
+public class DynamicRouteConfig {
```
Comment (manimaul) [inline xio-core/src/main/java/com/xjeffrose/xio/config/DynamicRouteConfig.java]: same comment as other pojos - lombok

File: xio-core/src/main/java/com/xjeffrose/xio/config/DynamicRouteConfigsFactory.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+ */
+public class DynamicRouteConfigsFactory {
+
+  /**
+   * This is a factory method used to invoke the build operation given a JSON input string The
+   * output of this file is the raw material used to build ProxyRoutes dynamically from JSON
+   */
+  public static ArrayList<DynamicRouteConfig> build(String string) {
+    Gson gson = new Gson();
+    DynamicRouteEntry[] dynamicRouteEntries = gson.fromJson(string, DynamicRouteEntry[].class);
```
Comment (manimaul) [inline xio-core/src/main/java/com/xjeffrose/xio/config/DynamicRouteConfigsFactory.java]: i'd prefer the use of a `List<DynamicRouteEntry>` over an array - works with both Gson and Jackson

File: xio-core/src/main/java/com/xjeffrose/xio/config/DynamicClientConfig.java (hunk 1 above)
Comment (khappucino) [reply xio-core/src/main/java/com/xjeffrose/xio/config/DynamicClientConfig.java]: ok applied that

File: xio-core/build.gradle (hunk 3 above)
Comment (khappucino) [reply xio-core/build.gradle]: ok swapping to Jackson

File: xio-core/src/main/java/com/xjeffrose/xio/config/DynamicClientConfig.java (hunk 4 above)
Comment (khappucino) [reply xio-core/src/main/java/com/xjeffrose/xio/config/DynamicClientConfig.java]: yup

File: xio-core/src/main/java/com/xjeffrose/xio/config/DynamicRouteConfig.java (hunk 5 above)
Comment (khappucino) [reply xio-core/src/main/java/com/xjeffrose/xio/config/DynamicRouteConfig.java]: yup

File: xio-core/src/main/java/com/xjeffrose/xio/config/DynamicRouteConfigsFactory.java (hunk 6 above)
Comment (khappucino) [reply xio-core/src/main/java/com/xjeffrose/xio/config/DynamicRouteConfigsFactory.java]: yeah i can swap that

File: xio-core/src/main/java/com/xjeffrose/xio/config/RouteReloader.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class RouteReloader<T> {
+
+  private static class RouteMeta<T> {
```
Comment (manimaul) [inline xio-core/src/main/java/com/xjeffrose/xio/config/RouteReloader.java]: static inner and inner classes usually are ordered to the bottom of the parent class

File: xio-core/src/main/java/com/xjeffrose/xio/config/RouteReloader.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+      digest.update(Files.readAllBytes(Paths.get(configFile.getAbsolutePath())));
+      return new RouteConfigFileMetadata(
+          configFile.getAbsolutePath(), configFile.lastModified(), digest.digest());
+    } catch (Exception e) {
+      throw new IllegalStateException(
+          "Couldn't load config from file '" + configFile.getAbsolutePath() + "'", e);
+    }
+  }
+
+  private boolean haveWatchFilesChanged() {
```
Comment (manimaul) [inline xio-core/src/main/java/com/xjeffrose/xio/config/RouteReloader.java]: this function is doing more than determining if the watch files have changed - it's also updating the files watched state
maybe a better name would be more appropriate or break apart

File: xio-core/src/test/java/com/xjeffrose/xio/config/RouteReloaderUnitTest.java [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+    TrivialState state =
+        new TrivialState() {
+          @Override
+          public void fireUpdated() {
+            fireUpdatedCalled.set(true);
+          }
+        };
+
+    subject.start(state::update);
+    Thread.sleep(5000);
```
Comment (manimaul) [inline xio-core/src/test/java/com/xjeffrose/xio/config/RouteReloaderUnitTest.java]: Thread sleeps are sad and slow. Consider an `@Test(timeout=...` and and observe when the subject is ready. You could possibly use a CountDownLatch as well.

File: xio-core/src/test/java/com/xjeffrose/xio/config/RouteReloaderUnitTest.java [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+    assertEquals(null, state.newValue);
+    executor.shutdown();
+  }
+
+  @Test
+  public void testReload_WhenWatchedFilesChange_Date_Was_Modified_and_Digest_Was_NOT_Changed()
+      throws Exception {
+
+    AtomicBoolean fireUpdatedCalled = new AtomicBoolean(false);
+    ScheduledExecutorService executor = new ScheduledThreadPoolExecutor(1);
```
Comment (manimaul) [inline xio-core/src/test/java/com/xjeffrose/xio/config/RouteReloaderUnitTest.java]: note: RxJava is available for use in tests which makes testing async scenarios simpler. See `ReverseProxyFunctionalTest` for some examples.

File: xio-core/src/main/java/com/xjeffrose/xio/config/RouteReloader.java (hunk 7 above)
Comment (khappucino) [reply xio-core/src/main/java/com/xjeffrose/xio/config/RouteReloader.java]: ok will move

File: xio-core/src/main/java/com/xjeffrose/xio/config/RouteReloader.java (hunk 8 above)
Comment (khappucino) [reply xio-core/src/main/java/com/xjeffrose/xio/config/RouteReloader.java]: ill rename it =p

File: xio-core/src/test/java/com/xjeffrose/xio/config/RouteReloaderUnitTest.java (hunk 9 above)
Comment (khappucino) [reply xio-core/src/test/java/com/xjeffrose/xio/config/RouteReloaderUnitTest.java]: i'll try test timeout

Comment (khappucino) [reply xio-core/src/test/java/com/xjeffrose/xio/config/RouteReloaderUnitTest.java]: I'll change this in  a later pull request

**Pull request merged**